### PR TITLE
[SelectionDAG] Add space-optimized forms of OPC_CheckPredicate

### PR DESCRIPTION
**Pull request by wangpc-pp**
We record the usage of each `Predicate` and sort them by usage.

For the top 8 `Predicate`s, we will emit a `PC_CheckPredicateN` to
save one byte.

Overall this reduces the llc binary size with all in-tree targets by
about 61K.

This is a recommit of 1a57927, which was reverted in bc98c31.

The CI failures occurred when doing expensive checks (with option
`LLVM_ENABLE_EXPENSIVE_CHECKS` being ON).

The key point here is that we need stable sorting result in the
test, but doing expensive checks uncovered the non-determinism of
`llvm::sort`. So `llvm::sort` is changed to `llvm::stable_sort`
in this revised patch.

And we use `llvm::MapVector` to keep insertion order.

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

**Review by metaflow — APPROVED** [review]
I am not an expert, thank you for fixing the test!

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 1]
```diff
@@ -88,6 +87,8 @@ class MatcherTableEmitter {
     DenseMap<const ComplexPattern *, unsigned> ComplexPatternUsage;
     // Record the usage of PatternPredicate.
     std::map<StringRef, unsigned> PatternPredicateUsage;
+    // Record the usage of Predicate.
+    DenseMap<TreePattern *, unsigned> PredicateUsage;
```
Comment (MaskRay) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:91]: This should use `MapVector`.  I've fixed `ComplexPatternUsage` in c230138011cbf07ad7caf9d256ae9d0c5032a974

**Review by MaskRay — APPROVED** [review]

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp (hunk 1 above)
Comment (wangpc-pp) [reply llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:91]: Thanks!

**Comment by ilovepi** [comment]
Oddly, I'm seeing a test failure locally in `TableGen/address-space-patfrags.td` that stops happening when I revert this patch? Any clues as to why we wouldn't see this on bots? There isn't anything non-standard in my CMake other than I've enabled `-DLLVM_REVERSE_ITERATION=On`. Well, that and assertions + the runtimes build, but this happens even if I just build and test llvm.

```
FAIL: LLVM :: TableGen/address-space-patfrags.td (36879 of 53624)
******************** TEST 'LLVM :: TableGen/address-space-patfrags.td' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /usr/local/google/home/paulkirth/llvm-fork/build/bin/llvm-tblgen -gen-dag-isel -I /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/../../include /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td 2>&1 | /usr/local/google/home/paulkirth/llvm-fork/build/bin/FileCheck -check-prefix=SDAG /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td
+ /usr/local/google/home/paulkirth/llvm-fork/build/bin/FileCheck -check-prefix=SDAG /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td
+ /usr/local/google/home/paulkirth/llvm-fork/build/bin/llvm-tblgen -gen-dag-isel -I /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/../../include /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td
/usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td:50:15: error: SDAG-NEXT: expected string not found in input
// SDAG-NEXT: // Predicate_pat_frag_b
              ^
<stdin>:183:11: note: scanning from here
 case 1: {
          ^
<stdin>:200:2: note: possible intended match here
 // Predicate_pat_frag_a
 ^

Input file: <stdin>
Check file: /usr/local/google/home/paulkirth/llvm-fork/llvm/test/TableGen/address-space-patfrags.td

-dump-input=help explains the following input dump.

Input was:
<<<<<<
           .
           .
           .
         178:  if (AddrSpace != 123 && AddrSpace != 455)
         179: return false;
         180: return true;
         181:
         182:  }
         183:  case 1: {
next:50'0               X error: no match found
         184:  // Predicate_truncstore
next:50'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
         185:  SDNode *N = Node;
next:50'0     ~~~~~~~~~~~~~~~~~~~
         186:  (void)N;
next:50'0     ~~~~~~~~~~
         187:  if (!cast<StoreSDNode>(N)->isTruncatingStore()) return false;
next:50'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         188: return true;
next:50'0     ~~~~~~~~~~~~~
           .
           .
           .
         195: if (cast<StoreSDNode>(N)->getAddressingMode() != ISD::UNINDEXED) return false;
next:50'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         196: return true;
next:50'0     ~~~~~~~~~~~~~
         197:
next:50'0     ~
         198:  }
next:50'0     ~~~
         199:  case 3: {
next:50'0     ~~~~~~~~~~~
         200:  // Predicate_pat_frag_a
next:50'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
next:50'1      ?                        possible intended match
         201:  SDNode *N = Node;
next:50'0     ~~~~~~~~~~~~~~~~~~~
         202:  (void)N;
next:50'0     ~~~~~~~~~~
         203: if (cast<MemSDNode>(N)->getAlign() < Align(2))
next:50'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         204: return false;
next:50'0     ~~~~~~~~~~~~~~
         205: return true;
next:50'0     ~~~~~~~~~~~~~
           .
           .
           .
>>>>>>

--

********************
********************
Failed Tests (1):
  LLVM :: TableGen/address-space-patfrags.td


Testing Time: 95.67s
```

**Comment by wangpc-pp** [comment]
> Oddly, I'm seeing a test failure locally in `TableGen/address-space-patfrags.td` that stops happening when I revert this patch? Any clues as to why we wouldn't see this on bots? There isn't anything non-standard in my CMake other than I've enabled `-DLLVM_REVERSE_ITERATION=On`. Well, that and assertions + the runtimes build, but this happens even if I just build and test llvm.

This can be reproduced on my local machine, can you file an issue please? This should be backported to LLVM 18.

**Comment by ilovepi** [comment]
> > Oddly, I'm seeing a test failure locally in `TableGen/address-space-patfrags.td` that stops happening when I revert this patch? Any clues as to why we wouldn't see this on bots? There isn't anything non-standard in my CMake other than I've enabled `-DLLVM_REVERSE_ITERATION=On`. Well, that and assertions + the runtimes build, but this happens even if I just build and test llvm.
> 
> This can be reproduced on my local machine, can you file an issue please? This should be backported to LLVM 18.

Thanks for taking a look, and fixing this. It looks like you've already started the backporting process in #79420.

**Pull request merged**